### PR TITLE
fix(sourcemap): using serde_json::to_string to quote sourcemap string

### PR DESCRIPTION
**Pull request by underfin**

### CODE REVIEW

File: crates/oxc_sourcemap/src/sourcemap.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ pub struct SourceMap {
     pub(crate) token_chunks: Option<Vec<TokenChunk>>,
 }
 
-#[allow(clippy::cast_possible_truncation)]
+#[allow(clippy::cast_possible_truncation, clippy::missing_errors_doc)]
```
Comment (Boshen) [inline crates/oxc_sourcemap/src/sourcemap.rs:20]: Let's document the error properly.

File: crates/oxc_sourcemap/src/encode.rs [hunk 2]
```diff
@@ -14,20 +15,35 @@ pub fn encode(sourcemap: &SourceMap) -> String {
...
+        .map(|x| serde_json::to_string(x.as_ref()))
+        .collect::<std::result::Result<Vec<_>, serde_json::Error>>()
+        .map_err(Error::from)?;
+    buf.push_str(&names.join(","));
     buf.push_str("],\"sources\":[");
-    buf.push_str(&sourcemap.sources.iter().map(|x| format!("{x:?}")).collect::<Vec<_>>().join(","));
+    let sources = sourcemap
+        .sources
+        .iter()
+        .map(|x| serde_json::to_string(x.as_ref()))
```
Comment (Boshen) [inline crates/oxc_sourcemap/src/encode.rs:29]: I think we should document why we need `serde_json::to_string`, and we should have a test case for this at least.

**Review by Boshen — COMMENTED** [review]
What's the reason for the performance improvement?

File: crates/oxc_sourcemap/src/sourcemap.rs (hunk 1 above)
Comment (underfin) [reply crates/oxc_sourcemap/src/sourcemap.rs]: Done

File: crates/oxc_sourcemap/src/encode.rs (hunk 2 above)
Comment (underfin) [reply crates/oxc_sourcemap/src/encode.rs:31]: Done

**Comment by underfin** [comment]
> What's the reason for the performance improvement?

I'm not found other different, here look like has a fast lookup table, https://github.com/serde-rs/json/commit/fc436ebd73e5bcafa24b34b3a3aeb44eede212a4

File: crates/oxc_sourcemap/src/encode.rs [hunk 3]
```diff
@@ -125,14 +143,24 @@ fn encode_vlq(out: &mut String, num: i64) {
...
 
     for (tok1, tok2) in sm.get_tokens().zip(sm2.get_tokens()) {
         assert_eq!(tok1, tok2);
     }
 }
+
+#[test]
+fn test_encode_escape_string() {
+    // '\0' should be escaped.
+    let sm = SourceMap::new(None, vec!["\0a.js".into()], vec![], None, vec![], None);
```
Comment (Boshen) [inline crates/oxc_sourcemap/src/encode.rs:161]: You mentioned "Here using `serde_json::to_string` to serialization `names/source_contents/sources`."

I think the test case should cover all 3.

**Review by Boshen — COMMENTED** [review]
I wonder whether we should bubble up the error or consume the error by replacing it with a placeholder such as "UnableAbleToSerializeString" into it, it feels weird that the consumer have to handle this error.

**Comment by underfin** [comment]
> I wonder whether we should bubble up the error or consume the error by replacing it with a placeholder such as "UnableAbleToSerializeString" into it, it feels weird that the consumer have to handle this error.

The `Error::BadJson(serde_json::Error)` already exported it.

**Comment by Boshen** [comment]
> > I wonder whether we should bubble up the error or consume the error by replacing it with a placeholder such as "UnableAbleToSerializeString" into it, it feels weird that the consumer have to handle this error.
> 
> The `Error::BadJson(serde_json::Error)` already exported it.

I mean the user of this crate need to handle this error, but I think they shouldn't.

**Comment by underfin** [comment]
Actually, the `serde_json` error does not help to resolve the issue, the message is not clear to json. I using the editor to see json problems.

**Comment by underfin** [comment]
> I mean the user of this crate need to handle this error, but I think they shouldn't.

It's hard to say, but the rolldown is already to handle this correctly.

**Review by Boshen — APPROVED** [review]

**Pull request merged**